src/connectors/queue/payTo/payToByBlockchain.ts

```diff
@@ -266,24 +180,386 @@ class PayToByBlockchainQueue extends BaseQueue {
+              targetId: article.id,
+            },
+            trx
+          )
+          await this.paymentService.baseUpdate(
+            blockchainTx.id,
+            {
+              transactionId: newTx.id,
+              state: BLOCKCHAIN_TRANSACTION_STATE.succeeded,
+            },
+            'blockchain_transaction',
+            trx
+          )
+          await trx.commit()
+        } catch (error) {
+          await trx.rollback()
+          throw error
+        }
+      }
+      this.notify({ tx, sender: curatorUser, recipient: creatorUser, article })
+      this.invalidCache(tx.targetType, tx.targetId)
+    } else {
+      // no related tx record, create one
+      const trx = await this.knex.transaction()
+      let tx
+      try {
+        tx = await this.paymentService.createTransaction(
+          {
+            amount,
+            state: TRANSACTION_STATE.succeeded,
+            purpose: TRANSACTION_PURPOSE.donation,
+            currency: PAYMENT_CURRENCY.USDT,
+            provider: PAYMENT_PROVIDER.blockchain,
+            providerTxId: blockchainTx.id,
+            recipientId: creatorUser.id,
+            senderId: curatorUser.id,
+            targetId: article.id,
+          },
+          trx
+        )
+        await this.paymentService.baseUpdate(
+          blockchainTx.id,
+          { transactionId: tx.id },
+          'blockchain_transaction',
+          trx
+        )
+        await trx.commit()
+      } catch (error) {
+        await trx.rollback()
+        throw error
+      }
+      this.notify({ tx, sender: curatorUser, recipient: creatorUser, article })
+      this.invalidCache(tx.targetType, tx.targetId)
+    }
+  }
+
+  private fetchCurationLogs = async (
+    curation: CurationContract,
+    savepoint: number | null
+  ): Promise<[Array<Log<CurationEvent>>, number]> => {
+    const safeBlockNum =
+      (await curation.fetchBlockNumber()) - BLOCKCHAIN_SAFE_CONFIRMS.Polygon
+
+    const fromBlockNum = savepoint ? savepoint + 1 : 0
+
+    if (fromBlockNum === 0) {
+      // no sync record in db , request getLog without block range
+      const logs = await curation.fetchLogs()
+      const filtered = logs.filter((e) => e.blockNumber <= safeBlockNum)
+
+      const newSavepoint =
+        logs.length === filtered.length
+          ? safeBlockNum
+          : filtered[filtered.length - 1].blockNumber
+
+      return [filtered, newSavepoint]
+    } else {
+      // sync record in db , request getLog with block range
+      // as provider only accept 2000 blocks range
+      const toBlockNum = Math.min(safeBlockNum, fromBlockNum + 1999)
+      if (fromBlockNum >= toBlockNum) {
+        return [[], savepoint as number]
+      }
+      return [await curation.fetchLogs(fromBlockNum, toBlockNum), toBlockNum]
+    }
+  }
+
+  private syncCurationEvents = async (logs: Array<Log<CurationEvent>>) => {
+    const events = []
+    for (const log of logs) {
+      if (!log.removed) {
```

Comment:
prefer to use [return-early pattern](https://dev.to/jpswade/return-early-12o5) to reduce nested code.

```
if (log.removed) {
    throw ...
}

const data: any = { ...log.event }
...
```